p2p/src/transport.rs

```diff
@@ -0,0 +1,129 @@
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+
+use crate::peer::{self, Peer};
+
+// TODO(xla): Use actual PeerId type.
+type PeerId = String;
+
+pub struct BindInfo {
+    pub addr: SocketAddr,
+    pub advertise_addrs: Vec<SocketAddr>,
+    pub peer_id: PeerId,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {}
+
+pub trait Connection: Clone {
+    type Stream;
+
+    fn advertised_addrs(&self) -> Vec<SocketAddr>;
+    fn close(&self) -> Result<(), Error>;
+    fn open(&self) -> Result<Self::Stream, Error>;
+    fn peer_id(&self) -> PeerId;
+    fn remote_addr(&self) -> SocketAddr;
+}
+
+pub trait Endpoint {
+    type Connection;
+
+    fn connect(&self) -> Result<Self::Connection, Error>;
+    fn listen_addrs(&self) -> SocketAddr;
+}
+
+pub trait Transport {
+    type Connection: Connection;
+    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
+    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection, Error>>;
```

Comment:
what about outgoing connections?

Reply:
We should definitely distinguish the direction of the connection. Maybe that is better done in the `Connection` type.

Reply:
Expressed in b5b0643